lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -149,24 +132,92 @@ impl std::fmt::Debug for OverlayContents {
     }
 }
 
-/// A kind of overlay page.
-#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq)]
+impl TryFrom<Vec<u8>> for OverlayContents {
+    type Error = usize;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        let len = value.len();
+        Ok(Self(Box::new(value.try_into().map_err(|_| len)?)))
+    }
+}
+
+/// A kind of overlay page, annotated with any other information that may be
+/// needed to generate the page's contents.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub(super) enum OverlayKind {
-    Hypercall,
+    /// A hypercall page whose instruction sequence immediately returns a "not
+    /// supported" error status to its caller.
+    HypercallReturnNotSupported,
 
+    /// A dummy page used for testing.
     #[cfg(test)]
-    Test(u8),
+    Test {
+        /// An index that disambiguates different test page kinds.
+        index: u8,
+
+        /// A byte with which to fill the overlaid page when this overlay
+        /// becomes active.
+        fill: u8,
+    },
+}
+
+impl OverlayKind {
+    /// Returns the "priority" of this page relative to overlay pages of other
+    /// kinds. For each PFN the page with the lowest priority value is active.
+    ///
+    /// WARNING: The priorities of existing overlay kinds should not be changed,
+    /// since this can cause a PFN's active overlay to change if a VM is
+    /// migrated from a Propolis using one ordering to a Propolis using a
+    /// different ordering.
```

Comment:
it occurs to me that we _could_ have a test that writes out the overlay kind priorities to a file and asserts the existing ones haven't changed, or something. not sure if this is worth bothering with.

Reply:
A guest stacking overlays is considered bad behavior, and the TLFS basically describes it similar to UB, where the observed result is largely up to the hypervisor.  I don't think we need to bend over backwards to provide a good experience for guests doing such things.